[VPlan] Dispatch to multiple exit blocks via middle blocks.

fhahn opened the pull request:
A more lightweight variant of https://github.com/llvm/llvm-project/pull/109193,
which dispatches to multiple exit blocks via the middle blocks.

The patch also introduces a bit of required scaffolding to enable early-exit vectorization, including an option. At the moment, early-exit vectorization doesn't come with legality checks, and is only used if the option is provided and the loop has metadata forcing vectorization. This is only intended to be used for testing during bring-up, with @david-arm enabling auto early-exit vectorization plugging in the changes from https://github.com/llvm/llvm-project/pull/88385.

fhahn committed:
[VPlan] Dispatch to multiple exit blocks via middle blocks.

A more lightweight variant of https://github.com/llvm/llvm-project/pull/109193,
which dispatches to multiple exit blocks via the middle blocks.

fhahn committed:
[VPlan] Support VPIRBBs and VPIRInst phis with multiple predecessors.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1696,3 +1702,79 @@ void VPlanTransforms::createInterleaveGroups(
+    BasicBlock *TrueSucc = ExitingTerm->getSuccessor(0);
+    BasicBlock *FalseSucc = ExitingTerm->getSuccessor(1);
+    VPIRBasicBlock *VPExitBlock;
+    if (OrigLoop->getUniqueExitBlock())
```
Again, this doesn't seem to interact well with loops that have countable early exits, which we already support vectorising. I will be adding support for loops with a mixture of countable and uncountable early exits (https://github.com/llvm/llvm-project/pull/88385), which will require a scalar epilogue for the final iteration (assuming we didn't leave via the uncountable exit). I imagine that's more efficient than testing the early countable exit in each vector iteration.

For the purposes of adding initial support for multiple loop region successors, you could also potentially simplify the algorithm by requiring `canVectorizeEarlyExit` to only permit one early exit? Or if you prefer to add support for multiple early exits then it makes sense to add tests for them. I think at the moment the tests only have a single, countable early exit.

david-arm replied:
Hmm, I think I understand what you're trying to do with this a bit more now. Essentially, this patch is trying to take an existing supported vectorisation case (early countable exits) and apply a new vectorisation style. This permits you to test the new code you're adding without having to worry about the legality of vectorising loops with uncountable exits. I guess it's just the new flag I'm worried about since it conflicts with my existing PR #88385. It's almost like you want a flag to control the flavour of vectorisation for an already support case, i.e. something like `-prefer-multi-exit`.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1514,6 +1522,27 @@ bool LoopVectorizationLegality::canVectorizeWithIfConvert() {
   return true;
 }
 
+bool LoopVectorizationLegality::canVectorizeEarlyExit() const {
```
It feels a little odd to have both `canVectorizeEarlyExit` and `isVectorizableEarlyExitLoop` in the same class. Would it make sense to move the hints check into `isVectorizableEarlyExitLoop` to bypass legality checks? You could even get the benefit of existing code to build up a list of countable and uncountable exits that can be used later on when splitting the middle.block.

david-arm replied:
I think you can also able enable early exit autovec with Transforms/LoopVectorize/single_early_exit.ll

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1696,3 +1702,79 @@ void VPlanTransforms::createInterleaveGroups(
+      BackedgeTakenCount, Plan.getCanonicalIV()->getScalarType(), OrigLoop);
+  VPValue *NewTC = vputils::getOrCreateVPValueForSCEVExpr(Plan, TripCount, SE);
+  Plan.getTripCount()->replaceAllUsesWith(NewTC);
+  Plan.resetTripCount(NewTC);
```
I don't think you need this code - the existing trip count already uses the predicated symbolic max backedge taken count.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1696,3 +1702,79 @@ void VPlanTransforms::createInterleaveGroups(
+        break;
+      Value *IncomingValue = ExitPhi->getIncomingValueForBlock(Exiting);
+      VPValue *V = RecipeBuilder.getVPValueOrAddLiveIn(IncomingValue);
+      ExitIRI->addOperand(V);
```
I think this is ok for now, but ultimately it will need handling in the same way as `collectUsersInExitBlock`. I have a downstream implementation of https://github.com/llvm/llvm-project/pull/88385 that avoids using VPLiveOut, etc. and refactors `collectUsersInExitBlock` to work for early exit blocks too. I'll put a patch up soon!

arcbbb replied:
Does the ExitIRI need an ExtractFromEnd if the vp value is not LiveIn?

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1696,3 +1702,79 @@ void VPlanTransforms::createInterleaveGroups(
   }
 }
+
+void VPlanTransforms::convertToMultiCond(VPlan &Plan, ScalarEvolution &SE,
```
It's not immediately obvious from the name what this means and is used for. I wonder if it's better named as something like `handleUncountableEarlyExit` to reflect what it will actually be used for? Or perhaps `handleMultipleRegionSuccessors`?

david-arm reviewed (commented):
Thanks for this! I realise you're busy with the LLVM dev conference, but thought I'd leave a few comments I have so far ...

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7692,12 +7697,15 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
-        dyn_cast<VPInstruction>(&R), State, OrigLoop,
-        State.CFG.VPBB2IRBB[ExitVPBB], ExpandedSCEVs);
+  VPBasicBlock *ExitVPBB = nullptr;
+  if (BestVPlan.getVectorLoopRegion()->getSingleSuccessor()) {
```
In fact there can only be one successor in this patch.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8778,6 +8789,8 @@ static void addCanonicalIVRecipes(VPlan &Plan, Type *IdxTy, bool HasNUW,
 static SetVector<VPIRInstruction *> collectUsersInExitBlock(
     Loop *OrigLoop, VPRecipeBuilder &Builder, VPlan &Plan,
     const MapVector<PHINode *, InductionDescriptor> &Inductions) {
+  if (!Plan.getVectorLoopRegion()->getSingleSuccessor())
```
There will only ever be one successor.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8863,6 +8876,8 @@ static void addLiveOutsForFirstOrderRecurrences(
   // TODO: Should be replaced by
   // Plan->getScalarLoopRegion()->getSinglePredecessor() in the future once the
   // scalar region is modeled as well.
+  if (!VectorRegion->getSingleSuccessor())
```
Only one successor permitted. Can delete this I think?

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9403,6 +9419,10 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
   }
   VPBasicBlock *LatchVPBB = VectorLoopRegion->getExitingBasicBlock();
   Builder.setInsertPoint(&*LatchVPBB->begin());
+  if (!VectorLoopRegion->getSingleSuccessor())
```
Again, only one successor.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -474,6 +474,14 @@ void VPIRBasicBlock::execute(VPTransformState *State) {
     // backedges. A backward successor is set when the branch is created.
     const auto &PredVPSuccessors = PredVPBB->getHierarchicalSuccessors();
     unsigned idx = PredVPSuccessors.front() == this ? 0 : 1;
+    if (TermBr->getSuccessor(idx) &&
```
I think this code was for a different version of the patch using BranchOnMultiCond - can be deleted I think?

david-arm commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1031,7 +1043,9 @@ void VPlan::execute(VPTransformState *State) {
   BasicBlock *MiddleBB = State->CFG.ExitBB;
   VPBasicBlock *MiddleVPBB =
-      cast<VPBasicBlock>(getVectorLoopRegion()->getSingleSuccessor());
+      getVectorLoopRegion()->getNumSuccessors() == 1
```
One successor only.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1056,12 +1074,19 @@ void VPlan::execute(VPTransformState *State) {
   VPBasicBlock *LatchVPBB = getVectorLoopRegion()->getExitingBasicBlock();
   BasicBlock *VectorLatchBB = State->CFG.VPBB2IRBB[LatchVPBB];
 
+  if (!getVectorLoopRegion()->getSingleSuccessor())
```
One successor only.

david-arm reviewed (commented):
Sorry @fhahn I left a few more comments for you to look at when you get back from the conference!

fhahn commented:
> Sorry @fhahn I left a few more comments for you to look at when you get back from the conference!

Great thanks, will get to them shortly, more to clean up after moving from https://github.com/llvm/llvm-project/pull/109193

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
!fixup address first set of comments, thanks!

fhahn reviewed (commented):
Addressed a first batch of comments, more to follow soon!

david-arm commented:
Hi @fhahn, just for reference here is a branch that contains the changes in this PR as a single initial commit, followed by 3 more commits to add support for early exit auto-vectorisation. It's just a reworking of my original PR #88385, based off this new approach.

https://github.com/david-arm/llvm-project/tree/ee_autovec2

Once this PR lands I'll update PR #88385 accordingly. Hopefully the branch will give you an indication of what direction I'm moving towards and might help to make sense of the comments I left on this PR too!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
!fixup clean up merge failures

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
[VPlan] Generalize collectUsersInExitBlocks for multiple exit bbs.

Generalize collectUsersInExitBlock to collecting exit users in multiple
exit blocks. Exit blocks are leaf nodes in the VPlan (without
successors) except the scalar header.

Split off from https://github.com/llvm/llvm-project/pull/112138

fhahn committed:
!fixup address comments

fhahn committed:
!fixup address more comments.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -43,6 +43,10 @@ AllowStridedPointerIVs("lv-strided-pointer-ivs", cl::init(false), cl::Hidden,
                        cl::desc("Enable recognition of non-constant strided "
                                 "pointer induction variables."));
 
+static cl::opt<bool>
```
Hmm, the flag still seems to be in the old place. Perhaps the patch hasn't updated correctly?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
!fixup remove left over canVectorizeEarlyExit

david-arm reviewed (commented):
Thanks for updating the patch and addressing all my previous comments! I have some more, but it's looking much better now. :)

david-arm commented:
Hi @fhahn, whilst doing some more testing of my patches based on this version I came across what looks like a serious issue in `collectUsersInExitBlocks`. One of the programs in the LLVM test suite (yacr2) miscompiles and I tracked the problem down to some invalid ordering assumptions in `VPIRInstruction::generate`. Essentially `collectUsersInExitBlocks` adds operands to the VPIRInstruction in the order of BasicBlock predecessors in the scalar loop, whereas `VPIRInstruction::generate` follows the order of VPBasicBlock predecessors. The assumption here, which is incorrect, is that the exiting VPBasicBlock in the predecessor list at index `I` corresponds like-for-like with the exiting BasicBlock predecessor index `I` in the original scalar loop. For yacr2 this isn't true and the incoming PHI values got swapped the wrong way around.

I've posted PR #117008, which shows how to fix this. That PR also does extra work required for enabling the tests, but you can look specifically at what I did for `VPIRInstruction::generate`. It's not particularly nice and you may have a better idea? Or is it worth scaling back this patch to only deal with the single early exit that's currently permitted by LoopVectorizationLegality? That might make life simpler and you can more easily enforce consistent ordering. What do you think?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
[VPlan] Print incoming VPBB for phi VPIRInstruction (NFC).

fhahn committed:
!fixup update recipe printing

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
!fixup add dbg message

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1375,6 +1375,10 @@ bool LoopVectorizationLegality::isFixedOrderRecurrence(
 bool LoopVectorizationLegality::blockNeedsPredication(BasicBlock *BB) const {
+  // When vectorizing early exits, create predicates for all blocks, except the
+  // header.
+  if (hasUncountableEarlyExit() && BB != TheLoop->getHeader())
```
Yeah given that this is now directly tied to what is supported by `hasUncountableEarlyExit` I think it makes sense to rely on the structure; updated, thanks

fhahn commented:
> Hi @fhahn, whilst doing some more testing of my patches based on this version I came across what looks like a serious issue in `collectUsersInExitBlocks`. One of the programs in the LLVM test suite (yacr2) miscompiles and I tracked the problem down to some invalid ordering assumptions in `VPIRInstruction::generate`. Essentially `collectUsersInExitBlocks` adds operands to the VPIRInstruction in the order of BasicBlock predecessors in the scalar loop, whereas `VPIRInstruction::generate` follows the order of VPBasicBlock predecessors. The assumption here, which is incorrect, is that the exiting VPBasicBlock in the predecessor list at index `I` corresponds like-for-like with the exiting BasicBlock predecessor index `I` in the original scalar loop. For yacr2 this isn't true and the incoming PHI values got swapped the wrong way around.

Thanks for checking that's a good find! I added a test with a swapped use order for the exit block to check for this case and also improved the printing to include which VPBB the value is coming from.

Given that the current logic is tied to the existing legality checks which only support a single early exit, I updated the code in `collectUsersInExitBlocks` to iterate over the predecessors in VPlan and distinguish between the latch and early exit by checking if the VPlan predecessor is the middle.bb. WDYT?

Also adde `-enable-early-exit-vectorization` to an existing test to cover end-to-end IR codegen and slightly updated the legality test to check for the new message. Not sure if this patch should add the flag to all related tests? I removed the earlier tests from this PR that checked codegen, as we already have existing tests for that.

fhahn committed:
!fixup fix formatting

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9965,12 +10000,14 @@ bool LoopVectorizePass::processLoop(Loop *L) {
+      reportVectorizationFailure("Auto-vectorization of loops with uncountable "
+                                 "early exit is not enabled",
+                                 "Auto-vectorization of loops with uncountable "
+                                 "early exit is no enabled",
```
nit: I think this should be `"early exit is not enabled"`

fhahn replied:
Should be fixed, thanks

david-arm commented on llvm/test/Transforms/LoopVectorize/X86/multi-exit-cost.ll:
Looks like these changes are not related to this patch?

fhahn replied:
It looks like those were just changing the pattern names, restored the original version, thanks

david-arm reviewed (commented):
LGTM! Thanks a lot for all the changes @fhahn. I have a couple of minor comments, but these can be addressed before merging.

david-arm commented:
Hi @fhahn, do know when you might be able to merge this patch? Thanks!

fhahn commented:
> 



> Hi @fhahn, do know when you might be able to merge this patch? Thanks!

Yep, planning on doing it either today or tomorrow, want to see if there are any further comments and clean up some of the unrelated changes

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
!restore test checks.

fhahn commented:
@david-arm just updated the patch on top of current main, but realized that it's not approved in Github yet. Would you mind pressing the button?

david-arm reviewed (approved):
LGTM! Sorry @fhahn, I obviously only left the LGTM comment before without actually approving it. :face_palm

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2601,9 +2607,9 @@ BasicBlock *InnerLoopVectorizer::emitMemRuntimeChecks(BasicBlock *Bypass) {
   assert(LoopVectorPreHeader && "Invalid loop structure");
-  assert((OrigLoop->getUniqueExitBlock() ||
+  assert((OrigLoop->getUniqueLatchExitBlock() ||
           Cost->requiresScalarEpilogue(VF.isVector())) &&
-         "multiple exit loop without required epilogue?");
+         "loops not exiting via the latch without required epilogue?");
```
Potentially committed separately?

david-arm replied:
If possible I'd prefer not to make this patch dependent on another PR, since this PR has already been blocking https://github.com/llvm/llvm-project/pull/88385 for two months. Could some of these comments be addressed after this PR lands or in an independent PR?

fhahn replied:
I'll split off the bits suggested by @ayalz and commit them separately without an extra PR given the changes have been approved and the benefit of splitting them off is to reduce the individual diffs (and make the changes clearly NFC)

david-arm replied:
Perfect, thank you!

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2841,6 +2845,8 @@ void InnerLoopVectorizer::fixupIVUsers(PHINode *OrigPhi,
     }
   }
 
+  assert((MissingVals.empty() || OrigLoop->getUniqueExitBlock()) &&
+         "Expected a single exit block for escaping values");
```
Potentially committed separately?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -3623,7 +3629,8 @@ void LoopVectorizationCostModel::collectLoopUniforms(ElementCount VF) {
   TheLoop->getExitingBlocks(Exiting);
   for (BasicBlock *E : Exiting) {
     auto *Cmp = dyn_cast<Instruction>(E->getTerminator()->getOperand(0));
-    if (Cmp && TheLoop->contains(Cmp) && Cmp->hasOneUse())
+    if (Cmp && TheLoop->contains(Cmp) && Cmp->hasOneUse() &&
+        (TheLoop->getLoopLatch() == E || !Legal->hasUncountableEarlyExit()))
```
Update above comment?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7796,15 +7803,17 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
   ILV.printDebugTracesAtEnd();
 
   // 4. Adjust branch weight of the branch in the middle block.
```
... "if the latter exists"?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8210,7 +8219,7 @@ VPValue *VPRecipeBuilder::createEdgeMask(BasicBlock *Src, BasicBlock *Dst) {
   // If source is an exiting block, we know the exit edge is dynamically dead
   // in the vector loop, and thus we don't need to restrict the mask.  Avoid
   // adding uses of an otherwise potentially dead instruction.
```
Update comment.

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8210,7 +8219,7 @@ VPValue *VPRecipeBuilder::createEdgeMask(BasicBlock *Src, BasicBlock *Dst) {
   // in the vector loop, and thus we don't need to restrict the mask.  Avoid
   // adding uses of an otherwise potentially dead instruction.
-  if (OrigLoop->isLoopExiting(Src))
+  if (!Legal->hasUncountableEarlyExit() && OrigLoop->isLoopExiting(Src))
```
Better check instead `if (OrigLoop->isLoopExiting(Src) && !(Src == Legal->getUncountableExitingBlock()))`?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9238,11 +9259,25 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
          "VPBasicBlock");
   RecipeBuilder.fixHeaderPhis();
 
+  if (Legal->hasUncountableEarlyExit()) {
+    assert(Legal->getUncountableExitingBlocks().size() == 1 &&
```
Easier to do `if (auto *UncountableExitingBlock = Legal->getUncountableExitingBlock())`?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9238,11 +9259,25 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
+        RecipeBuilder);
+  }
   addScalarResumePhis(RecipeBuilder, *Plan);
   SetVector<VPIRInstruction *> ExitUsersToFix = collectUsersInExitBlocks(
       OrigLoop, RecipeBuilder, *Plan, Legal->getInductionVars());
```
Better have `collectUsersInExitBlocks()` indicate failure, than `addUsersInExitBlocks()`, which follows `addExitUsersForFirstOrderRecurrences()`?

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9996,12 +10031,14 @@ bool LoopVectorizePass::processLoop(Loop *L) {
-                               "early exit is not yet supported",
-                               "UncountableEarlyExitLoopsUnsupported", ORE, L);
-    return false;
+    if (!EnableEarlyExitVectorization) {
```
```suggestion
    if (LVL.hasUncountableEarlyExit() && !EnableEarlyExitVectorization) {
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -860,14 +860,10 @@ VPlanPtr VPlan::createInitialVPlan(Type *InductionTy,
-  // getSymbolicMaxBackedgeTakenCount allows enablement work for loops with
-  // uncountable exits whilst also ensuring the symbolic maximum and known
-  // back-edge taken count remain identical for loops with countable exits.
+  // We use the symbolic max backedge-taken-count, which is used when
```
```suggestion
  // We use the symbolic max backedge-taken-count, which works also when
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -902,9 +898,10 @@ VPlanPtr VPlan::createInitialVPlan(Type *InductionTy,
+  BasicBlock *IRExitBlock = TheLoop->getUniqueLatchExitBlock();
   auto *VPExitBlock = VPIRBasicBlock::fromBasicBlock(IRExitBlock);
-  // The connection order corresponds to the operands of the conditional branch.
+  // The connection order corresponds to the operands of the conditional
+  // branch.
```
nit: unrelated.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1069,7 +1066,10 @@ void VPlan::execute(VPTransformState *State) {
       // Move the last step to the end of the latch block. This ensures
       // consistent placement of all induction updates.
       Instruction *Inc = cast<Instruction>(Phi->getIncomingValue(1));
-      Inc->moveBefore(VectorLatchBB->getTerminator()->getPrevNode());
+      if (VectorLatchBB->getTerminator() == &*VectorLatchBB->getFirstNonPHI())
+        Inc->moveBefore(VectorLatchBB->getTerminator());
+      else
+        Inc->moveBefore(VectorLatchBB->getTerminator()->getPrevNode());
```
Worth extending the comment?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1230,6 +1230,7 @@ class VPInstruction : public VPRecipeWithIRFlags,
     // operand). Only generates scalar values (either for the first lane only or
     // for all lanes, depending on its uses).
     PtrAdd,
+    AnyOf,
```
Some explanation how AnyOf relates (or should relate) to ComputeReductionResult?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3884,10 +3887,10 @@ class VPlan {
   /// latch.
   const VPBasicBlock *getMiddleBlock() const {
-    return cast<VPBasicBlock>(getVectorLoopRegion()->getSingleSuccessor());
+    return cast<VPBasicBlock>(getScalarPreheader()->getSinglePredecessor());
```
This restricts to use of getMiddleBlock() to before bypassing guards are introduced?

ayalz replied:
Works as long as the scalar preheader has a single predecessor, i.e., until runtime guards are introduced as additional predecessors.

fhahn replied:
Yep, this will need some extra work for https://github.com/llvm/llvm-project/pull/114292, which I plan to land after this PR.

ayalz reviewed (commented):
Post-approval review comments (potentially to be continued).

Should the block diagram of https://llvm.org/docs/Vectorizers.html#epilogue-vectorization be updated?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn committed:
!fixup address most comment, a few more pending

ayalz reviewed (commented):
Post-approval review comments (continued).

fhahn committed:
!fixup address remaining comments, thanks!

fhahn committed:
!fixup update doc

fhahn reviewed (commented):
> Should the block diagram of https://llvm.org/docs/Vectorizers.html#epilogue-vectorization be updated?

Added a brief section with a diagram of the VPlan for loops with early exits, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1820,6 +1820,57 @@ void VPlanTransforms::createInterleaveGroups(
+  }
+
+  VPValue *M = RecipeBuilder.getBlockInMask(
+      OrigLoop->contains(TrueSucc) ? TrueSucc : FalseSucc);
```
Also refers to the fact that only a single block is being masked in this case - the latch.

ayalz commented on llvm/test/Transforms/LoopVectorize/single_early_exit.ll:
```diff
@@ -119,22 +171,66 @@ define i32 @diff_exit_block_needs_scev_check(i32 %end) {
+; CHECK-NEXT:    [[TMP13:%.*]] = icmp eq <4 x i32> [[WIDE_LOAD]], [[WIDE_LOAD3]]
+; CHECK-NEXT:    [[TMP14:%.*]] = xor <4 x i1> [[TMP13]], splat (i1 true)
+; CHECK-NEXT:    [[INDEX_NEXT]] = add nuw i64 [[INDEX]], 4
+; CHECK-NEXT:    [[TMP15:%.*]] = xor <4 x i1> [[TMP14]], splat (i1 true)
```
Your call.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1838,3 +1838,61 @@ void VPlanTransforms::convertToConcreteRecipes(VPlan &Plan) {
+  auto *IsLatchExitTaken =
+      Builder.createICmp(CmpInst::ICMP_EQ, LatchExitingBranch->getOperand(0),
+                         LatchExitingBranch->getOperand(1));
+  auto *AnyExiting = Builder.createNaryOp(Instruction::Or,
```
```suggestion
  auto *AnyExitTaken = Builder.createNaryOp(Instruction::Or,
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1838,3 +1838,61 @@ void VPlanTransforms::convertToConcreteRecipes(VPlan &Plan) {
+  // with one exiting if either the original condition of the vector latch is
+  // true or the early exit has been taken.
+  auto *LatchExitingBranch =
+      dyn_cast<VPInstruction>(LatchVPBB->getTerminator());
```
Also good to assert the opcode of LatchExitingBranch, before taking its operands.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-branch-on-multi-cond-in-middle

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.h:
```diff
@@ -124,6 +124,17 @@ struct VPlanTransforms {
+  /// UncountableExitingBlock by
+  ///  * updating the condition exiting the vector loop to include the early
+  ///    exit conditions
+  ///  * splitting the original middle block to branch to the early exit blocks
```
updated thanks

fhahn committed:
!fixup address latest comments, thanks!

david-arm commented:
> Updated, it required updating getUncountableEarlyExitingBlock to return nullptr in if !UncountableEarlyExit and some slight adjustments on how it is set (and clear the entries if it is false).
> 
> WDYT @david-arm ?

Sorry I only saw this now. It was hard to see amongst all the other comments. The original idea behind `getUncountableEarlyExitingBlock` was that you should only be calling this if you know you have an uncountable early exit. That's what the existing assert

```
    assert(getUncountableExitingBlocks().size() == 1 &&
           "Expected only a single uncountable exiting block");
```

was doing. It's pretty difficult to parse all of the comment history in github - is the reason for the change because you want to avoid calling `hasUncountableEarlyExit` and are doing this as a shortcut? That seems fine.

fhahn commented:
> was doing. It's pretty difficult to parse all of the comment history in github - is the reason for the change because you want to avoid calling `hasUncountableEarlyExit` and are doing this as a shortcut? That seems fine.

Yep exactly, it is in some places more convenient to have a single call that returns `nullptr` if there's no vectorizable early exit block

david-arm commented:
Many thanks for all your work on this patch @fhahn!
